src/types/sgrStack.cpp

```diff
@@ -0,0 +1,240 @@
+    {
+    }
+
+    void SgrStack::Push(const TextAttribute& currentAttributes,
+                        const VTParameters options) noexcept
+    {
+        AttrBitset validParts;
+
+        try
+        {
+            if (options.empty())
+            {
+                // We save all current attributes.
+                validParts.set(static_cast<size_t>(SgrSaveRestoreStackOptions::All));
+            }
+            else
+            {
+                // Each option is encoded as a bit in validParts. All options (that fit) are
+                // encoded; options that aren't supported are ignored when read back (popped).
+                // So if you try to save only unsupported aspects of the current text
+                // attributes, you'll do what is effectively an "empty" push (the subsequent
+                // pop will not change the current attributes), which is the correct behavior.
+
+                for (size_t i = 0; i < options.size(); i++)
+                {
+                    // TRICKY: The VTParameter could be a fake/materialized parameter, in
+                    // which case it has a default value, and the default value of a
+                    // VTParameter depends on what you are assigning to: if you assign /
+                    // convert directly to a size_t, then the default is "1", but for other
+                    // size_t-sized types, the default is "0". We want "default is 0", so we
+                    // convert to a SgrSaveRestoreStackOptions first.
+                    const SgrSaveRestoreStackOptions option = options.at(i);
+                    const size_t optionAsIndex = static_cast<size_t>(option);
+
+                    // Options must be specified singly; not in combination. Values that are
+                    // out of range will be ignored.
+                    if (optionAsIndex < validParts.size())
+                    {
+                        validParts.set(optionAsIndex);
+                    }
+                }
+            }
+        }
+        catch (std::out_of_range&)
+        {
+            // We should not be able to reach here: we pre-check that everything should be
+            // in range.
+            RaiseFailFastException(nullptr, nullptr, 0);
+        }
+
+        if (_numSavedAttrs < gsl::narrow<int>(_storedSgrAttributes.size()))
+        {
+            _numSavedAttrs++;
+        }
+
+        _storedSgrAttributes.at(_nextPushIndex) = { currentAttributes, validParts };
+        _nextPushIndex = (_nextPushIndex + 1) % gsl::narrow<int>(_storedSgrAttributes.size());
+    }
+
+    const TextAttribute SgrStack::Pop(const TextAttribute& currentAttributes) noexcept
+    {
+        if (_numSavedAttrs > 0)
+        {
+            _numSavedAttrs--;
+
+            if (_nextPushIndex == 0)
+            {
+                _nextPushIndex = gsl::narrow<int>(_storedSgrAttributes.size() - 1);
+            }
+            else
+            {
+                _nextPushIndex--;
+            }

```

Comment:
nit: technically, you can do a wraparound mod math trick here:

```cpp
_nextPushIndex = (size + _nextPushIndex - 1) % size
```

That's if concision was the goal, of course. It's not necessarily the goal!

Reply:
Hmmm... I waffled on this. I lean toward the more verbose, but I'd change it if y'all had a strong opinion.